src/olympia/reviewers/management/commands/auto_reject.py

```diff
@@ -97,46 +103,68 @@ def reject_versions(self, *, addon, versions, latest_version):
+        # trigger the corresponding tasks.
+        _discard_tasks()
+        # Queue celery tasks for this version, avoiding triggering them too
+        # soon...
+        _start_queuing_tasks()
+
         try:
-            self.reject_versions(
-                addon=addon, versions=versions, latest_version=latest_version
-            )
+            with transaction.atomic():
+                # ...and release the queued tasks to celery once transaction
+                # is committed.
+                transaction.on_commit(_send_tasks_and_stop_queuing)
+                latest_version = addon.find_latest_version(channel=amo.CHANNEL_LISTED)
+                if (
+                    latest_version
+                    and latest_version.is_unreviewed
+                    and not latest_version.pending_rejection
+                ):
+                    # If latest version is unreviewed and not pending
+                    # rejection, we want to put the delayed rejection of all
+                    # versions of this addon on hold until a decision has been
+                    # made by reviewers on the latest one.
+                    log.info(
+                        'Skipping rejections for add-on %s until version %s '
+                        'has been reviewed',
+                        addon.pk,
+                        latest_version.pk,
+                    )
+                    return
+                versions = self.fetch_version_candidates_for_addon(addon=addon, now=now)
+                if not versions.exists():
+                    log.info(
+                        'Somehow no versions to auto-reject for add-on %s', addon.pk
+                    )
+                    return
+                locked_by = get_reviewing_cache(addon.pk)
+                if locked_by:
+                    # Don't auto-reject something that has been locked, even by the
+                    # task user - wait until it's free to avoid any conflicts.
+                    log.info(
+                        'Skipping rejections for add-on %s until lock from user %s '
+                        'has expired',
+                        addon.pk,
+                        locked_by,
+                    )
+                    return
+                set_reviewing_cache(addon.pk, settings.TASK_USER_ID)
+                self.reject_versions(
+                    addon=addon, versions=versions, latest_version=latest_version
+                )
         finally:
             # Always clear our lock no matter what happens.
             clear_reviewing_cache(addon.pk)
+            # Stop post request task queue before moving on (useful in tests to
+            # leave a fresh state for the next test. Note that we don't want to
+            # send or clear queued tasks (they may belong to a transaction that
+            # has been rolled back, or they may not have been processed by the
+            # on commit handler yet).
```

Comment:
As in `auto_approve.py` above that comment is no longer accurate now that we're also discarding any tasks left in the queue.

Reply:
https://github.com/mozilla/addons-server/pull/22305/commits/a2f16dbb07048d6fde39ce50b2547805cc9199fb